src/simplify_reshapes.cpp

```diff
@@ -86,25 +88,59 @@ bool is_no_transpose(const std::vector<int64_t>& dims)
                dims.begin(), dims.end(), [](auto x, auto y) { return (y - x) != 1; }) == dims.end();
 }
 
-struct find_reshaper
+struct find_nested_shape_transforms
 {
+    static const auto& shape_transforms()
+    {
+        static const std::unordered_set<std::string> names = {
+            "flatten",
+            "reshape",
+            "squeeze",
+            "unsqueeze",
+            "transpose",
+            "broadcast",
+            "multibroadcast",
+        };
+        return names;
+    }
     auto matcher() const
     {
-        auto reshaper          = match::name(reshaper_names());
-        auto contiguous        = match::name("contiguous");
-        auto no_output_reshape = match::none_of[match::outputs()](reshaper);
-        auto input_reshape     = match::arg(0)(match::skip(contiguous)(reshaper));
-        auto input             = match::skip(reshaper, contiguous)(match::any().bind("x"));
-        return reshaper(no_output_reshape, input_reshape, input);
+        auto output_not_shape_transform = match::none_of(
+            match::skip_output(match::name("contiguous"))(match::name(shape_transforms())));
+        auto input_has_shape_transform =
+            match::args(match::skip(match::name("contiguous"))(match::name(shape_transforms())));
+        return match::name(shape_transforms())(output_not_shape_transform,
+                                               input_has_shape_transform);
     }
 
     void apply(module& m, const match::matcher_result& mr) const
     {
-        auto ins   = mr.result;
-        auto input = mr.instructions["x"];
-        auto dims  = ins->get_shape().lens();
+        auto ins = mr.result;
 
-        m.replace_instruction(ins, make_op("reshape", {{"dims", dims}}), input);
+        std::vector<operation> ops;
+        auto x = ins;
+        while(contains(shape_transforms(), x->get_operator().name()) or
+              x->get_operator().name() == "contiguous")
+        {
+            ops.push_back(x->get_operator());
+            x = x->inputs().front();
+        }
+        if(x->get_shape().scalar())
+        {
+            m.replace_instruction(
+                ins, make_op("multibroadcast", {{"out_lens", ins->get_shape().lens()}}), x);
+        }
```

Comment:
I think this bugs out if you have a broadcasted scalar. Not sure if it's possible to get a broadcasted scalar with no shape transform instructions before it.

Reply:
If there is no shape transform instructions then this is not matched. There has to be at least two shape transforms for this to match.